### ISSUE TITLE
[BUG] Index out of bounds Error when passing array as Query if its not first parameter

### ISSUE THREAD

**Issue by moliver-bb** [reactiveui/refit#712]
**Describe the bug**
`
[Get("/v1/project/{projectId}/markups")]
Task<List<Users>> GetMarkups(string projectId, [Query(CollectionFormat.Multi)] string[] status);
`
fails with:
`
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: index
   at System.Linq.Enumerable.ElementAt[TSource](IEnumerable`1 source, Int32 index)
   at Refit.RequestBuilderImplementation.<>c__DisplayClass17_0.<<BuildRequestFactoryForMethod>b__0>d.MoveNext() in d:\a\1\s\Refit\RequestBuilderImplementation.cs:line 569
--- End of stack trace from previous location where exception was thrown ---
   at Refit.RequestBuilderImplementation.<>c__DisplayClass14_0`2.<<BuildCancellableTaskFuncForMethod>b__0>d.MoveNext() in d:\a\1\s\Refit\RequestBuilderImplementation.cs:line 244
--- End of stack trace from previous location where exception was thrown ---
`

**Steps To Reproduce**
Pass any array as Query where it is not the first parameter and it will throw. When passed as the first argument everything works fine. This works fine:
`
[Get("/v1/project/{projectId}/markups")]
Task<List<Users>> GetMarkups([Query(CollectionFormat.Multi)] string[] status, string projectId);
`

**Expected behavior**
Shouldn't throw an error should work as argument in any position

**Environment**

- OS:      dotnet core 2.2 on Mac
- Version   4.7.40
- Device:   NA

**Comment by benjaminhowarth1** [comment]
I'll take this one.

**Comment by DOMZE** [comment]
Hello,

Having the same problem when using the multi parameter in the Query attribute

Here's the signature of the problematic method
`[Get("/api/v1/user/info")]
Task<IEnumerable<UserInformation>> GetUserInformationAsync([Header("Authorization")] string authorization, [Query(CollectionFormat.Multi)][AliasAs("userId")]IEnumerable<string> userIds);`

However this works
`[Get("/api/v1/user/info")]
Task<IEnumerable<UserInformation>> GetUserInformationAsync([Query(CollectionFormat.Multi)][AliasAs("userId")]string[] userIds, [Header("Authorization")] string authorization);`

**Comment by benjaminhowarth1** [comment]
@DOMZE @moliver-bb can you both try and reproduce in the latest Refit that just got published to Nuget, 4.7.51?

**Comment by CreepyGnome** [comment]
I am getting this same exception as of today but we don't use Query attribute or Query strings at all in any of our requests. Everything worked yesterday and today this just started occurring with 4.7.9 and updating to 4.7.51 didn't fix it.  I would have to assume this is a conflict with something environmental and I checked all my other packages and only one changes since yesterday so I reverted it back one version and still no good. 

I when through and cannot find any changes to the interface, Refit client setup or the code where the call is made via the refit client this code has been unchanged for over a week and it was working yesterday just fine. 

I checked Windows update and I have not had any updates applied in the last 24 hours.

The only environmental change that I cannot role back to verify easily is that this morning there was a Visual Studio 2019 update which took me from 16.2.2 to 16.2.3. 

Could a Visual Studio update have broken Refit?  I couldn't imagine that occurring. I am going to start going grabbing at straws and doing some more digging. This is very odd behavior for a library to break with no known changes to its environment. 


This is the signature:
```csharp
[Get("/Read")]

Task<SaveResult> ReadSettings([Header(HeaderNames.Authorization)] string authorization, long subscriberId, int profileId, ChannelType channelType = ChannelType.Store, CancellationToken cancellationToken = default(CancellationToken));
```
Exception:
```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: index
   at TSource System.Linq.Enumerable.ElementAt<TSource>(IEnumerable<TSource> source, int index)
   at void Refit.RequestBuilderImplementation+<>c__DisplayClass17_0+<<BuildRequestFactoryForMethod>b__0>d.MoveNext() in d:/a/1/s/Refit/RequestBuilderImplementation.cs:line 594
   at void Refit.RequestBuilderImplementation+<>c__DisplayClass14_0<T, TBody>+<<BuildCancellableTaskFuncForMethod>b__0>d.MoveNext() in d:/a/1/s/Refit/RequestBuilderImplementation.cs:line 244
   at async Task<ActionResult<ChannelSettings>> [REMOVED].SettingsController.Read(long subscriberId, int profileId, ChannelType channelType) in [REMOVED]/Controllers/SettingsController.cs:line 185
   at object lambda_method(Closure, object)
   at object Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable+Awaiter.GetResult()
   at async ValueTask<IActionResult> Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor+AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, object controller, object[] arguments)
   at TResult System.Threading.Tasks.ValueTask<TResult>.get_Result()
   at async Task Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionMethodAsync()
   at async Task Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()
   at void Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)
   at Task Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
   at async Task Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
   at async Task Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
   at void Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   at Task Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
   at async Task Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
   at async Task Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
   at async Task Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)
   at async Task Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.Invoke(HttpContext httpContext)
   at async Task Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at async Task Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at async Task Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at async Task Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at async Task Serilog.AspNetCore.RequestLoggingMiddleware.Invoke(HttpContext httpContext)
   at async Task Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.Invoke(HttpContext context)
```

**Comment by CreepyGnome** [comment]
Update: Another developer who still had VIsual Studio 2019 16.2.2 reproduces the issue on his box as well. So this is multiple environments now leaving the changes have to be within the solution itself, and less likely being outside of it.

**Comment by john-manktelow** [comment]
I'm getting the same exception, but only since 4.7.51. Changing the order of the parameters can work around it, though. For me, it's having a `[Header]` parameter first that triggers it.

**Repro**

```
using System;
using System.Net.Http;
using System.Threading.Tasks;
using Refit;
using RichardSzalay.MockHttp;

namespace refitRepro
{
    public interface ITheService
    {
        // Works
        [Get("/v1/HeaderSecond")]
        Task<string> HeaderSecond(string term, [Header("Some-Header")]string someHeader);

        // Breaks
        [Get("/v1/HeaderFirst")]
        Task<string> HeaderFirst([Header("Some-Header")]string someHeader, string term);
    }
    
    class Program
    {
        static void Main(string[] args)
        {
            const string rootUrl = "http://localhost";

            var mockHttp = new MockHttpMessageHandler();

            mockHttp.Expect(HttpMethod.Get, rootUrl + "/v1/HeaderSecond")
                .WithHeaders("Some-Header", "someHeader")
                .Respond("application/json", "result");
            
            mockHttp.Expect(HttpMethod.Get, rootUrl + "/v1/HeaderFirst")
                .WithHeaders("Some-Header", "someHeader")
                .Respond("application/json", "result");
            
            var sut = RestService.For<ITheService>(rootUrl,
                new RefitSettings() {HttpMessageHandlerFactory = () => mockHttp});

            
            // If the header is not the first param, everything is fine.
            var taskWorks = sut.HeaderSecond("headerSecond", "someHeader");
            Task.WaitAll(taskWorks);
            Console.WriteLine("Got a result when the header is the second param.");

            // If the header is the first param, we get an ArgumentOutOfRangeException from the RequestBuilderImplementation
            var taskBreaks = sut.HeaderFirst("someHeader", "headerFirst");
            Task.WaitAll(taskBreaks);
            Console.WriteLine("Exception thrown if the header is the first param.");
        }
    }
}

```
**Exception**
```
Unhandled exception. System.AggregateException: One or more errors occurred. (Specified argument was out of the range of valid values. (Parameter 'index'))
 ---> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'index')
   at System.Linq.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument)
   at System.Linq.Enumerable.ElementAt[TSource](IEnumerable`1 source, Int32 index)
   at Refit.RequestBuilderImplementation.<>c__DisplayClass17_0.<<BuildRequestFactoryForMethod>b__0>d.MoveNext() in d:\a\1\s\Refit\RequestBuilderImplementation.cs:line 594
--- End of stack trace from previous location where exception was thrown ---
   at Refit.RequestBuilderImplementation.<>c__DisplayClass14_0`2.<<BuildCancellableTaskFuncForMethod>b__0>d.MoveNext() in d:\a\1\s\Refit\RequestBuilderImplementation.cs:line 244
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.WaitAllCore(Task[] tasks, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.WaitAll(Task[] tasks)
   at refitRepro.Program.Main(String[] args) in Program.cs:line 47

```

**Workaround**
Change the order of the parameters, or revert to Refit 4.7.9

**Comment by martincostello** [comment]
Same problem here with 4.7.51. This interface worked in 4.7.9 but throws `ArgumentOutOfRangeException` in 4.7.51:

```csharp
internal interface IFoo
{
    [Put("/foo/{id1}/bar")]
    Task<ApiResponse<string>> DoThingAsync(
        [Header("x-custom-header")] string myHeader,
        string id1,
        string id2,
        JObject body,
        CancellationToken cancellationToken);
}
```

```
   at System.Linq.Enumerable.ElementAt[TSource](IEnumerable`1 source, Int32 index)
   at Refit.RequestBuilderImplementation.<>c__DisplayClass17_0.<<BuildRequestFactoryForMethod>b__0>d.MoveNext() in d:\a\1\s\Refit\RequestBuilderImplementation.cs:line 594
--- End of stack trace from previous location where exception was thrown ---
   at Refit.RequestBuilderImplementation.<>c__DisplayClass14_0`2.<<BuildCancellableTaskFuncForMethod>b__0>d.MoveNext() in d:\a\1\s\Refit\RequestBuilderImplementation.cs:line 244
```

I just noticed that it has an extra parameter left over from a previous version of code that's no longer used, which might be contributing to the problem.

**Comment by martincostello** [comment]
Explicitly adding the query string parameter to the URL template works around the issue.

**Comment by benjaminhowarth1** [comment]
I think that because we added POCO support to [Query] params, the IndexOf function got modified, cause previously if you had an IEnumerable on the POCO, it threw index errors (if I recall my debugging sessions correctly).

I'll pick this up & look to try and fix before Tuesday, it's a 3-day weekend here in the UK.

**Comment by onovotny** [comment]
Thanks. The CI feed is now on Azure Artifacts now btw. 

Let’s ensure we have test cases that cover these scenarios too.

**Comment by benjaminhowarth1** [comment]
@onovotny picking this up today alongside #696 and #587, hopefully I can get a PR in by tomorrow

**Comment by CreepyGnome** [comment]
@benjaminhowarth1 did you happen to make any progress on this?

**Comment by CreepyGnome** [comment]
Just additional confirmation, that for me at least, in 4.7.51 it is the header attribute being first and moving it to after query string values it works, and making no changes and rolling back to 4.7.9 works.

I rolled back to 4.7.9 as this is a breaking change in a patch version which should never occur. Also I would argue this type of breaking change should happen in a minor version as well.  Even a major version breaking change would be bad for this impact unless there was good justification for it.

Questions for those who maintain and release refit: 

1. Why are there no release notes for 4.7.51? 
2. Why was 4.7.51 released, what are the important changes?
3. Are there no tests around this functionality that triggered the breaking change?

I would recommend the release should be _removed from GitHub_ and _unlisting in nuget_ as this is is a huge breaking change. It killed just about all my refit calls and there is no way I am going to go back and refactor all of them to move the auth header from first to after the query string values.

**Comment by fouadroumieh** [comment]
This issue is not fixed in 4.7.51 and its quite critical issue, my first time using refit and its quite annoying!

**Comment by onovotny** [comment]
@benjaminhowarth1 any updates here?

**Comment by stoff99** [comment]
Hey @onovotny,

this is still not fixed. I get the same with the latest 5.0.15. Here is my api definition:

[Headers("Accept:application/json", "X-API-V: 125")]
[Get("/api/someModule/deviceList?controlId={control_id}")]
Task<DeviceListResponse> DeviceList([Header("Authorization")] string authorization, [Header("X-Lng")] string twoLetterLang, string search, [AliasAs("control_id")] string controlId, string secret);

**Comment by onovotny** [comment]
This should be finally fixed in 5.0.23, rolling out to NuGet now. If it crops up again, please open a new issue. The interface definitions have been helpful to validate these fixes.

**Comment by CreepyGnome** [comment]
@onovotny So it was broken in a patch update, but is being fixed in a major version update? Is there any way to fix it in at least a minor/patch so more people can upgrade to it if they have change controls going to major versions?  

We had to roll back to the last working version (4.7.9) of refit as we have too many interfaces using refit and would be a large undertaking to refactor all that, and then do a full regression.  We cannot just go from a 4.x to a 5.x without a lot of additional work and a full regression test because of a major version update in a library. This fix, as-is, will still be out of our reach in several of our larger and older projects.

**Comment by onovotny** [comment]
Refit has a policy of always taking the latest stable version version of its dependencies as they are available. That forced a major version bump to be in-line with SemVer. That said, we generally only release off of master and fix-forward.

**Comment by CreepyGnome** [comment]
I hope you understand the problem with that policy.  

That a hotfix policy may be appropriate when you release an unexpected breaking change in a patch release so that you can fix/revert what is needed to release another patch release to address what was broken. Preferably quickly.  Hotfix branches and CICD release pipelines are something that could be implemented to make it easy to manage a hotfix release.

**Comment by stoff99** [comment]
@onovotny the bug is fixed with 5.0.23 now.
Thx a lot!
